### ISSUE TITLE
Navigation profile seems never correctly initialized for GPX-based routes

### ISSUE THREAD

**Issue by sonora** [osmandapp/OsmAnd#15308]
Let me document this as an issue so it doesn't get lost:

After using the new _NavigateNotSavedTrack_ feature in the latest nightly (and if this is the first navigation you try after last starting OsmAnd), the navigation profiles get messed up somehow,  e.g. the user elevation profile option buttons disappear for bike or ped routing. The RouteOptionsBottomSheet entries seem to be only those of the OTHER case of https://github.com/osmandapp/OsmAnd/blob/22e40f113ce5c6df97f2f1687d5024ae38a4d28b/OsmAnd/src/net/osmand/plus/routepreparationmenu/RouteOptionsBottomSheet.java#L863 This can only be fixed by killing the app and restarting.

https://github.com/osmandapp/OsmAnd/commit/37994147e4a66005f4bd9fc50bc0101ed8efcdcb#commitcomment-83927469
https://github.com/osmandapp/OsmAnd/commit/37994147e4a66005f4bd9fc50bc0101ed8efcdcb#commitcomment-83928662

![image](https://user-images.githubusercontent.com/845735/190212564-54d93761-5c09-4449-b97a-a858b13e44c1.png)

**Comment by sonora** [comment]
I just notice that this issues is not limited to using a "not-saved" route you tapped on the fly:

On the route preparation menu the buttons "Use elevation" for the ped profile or in addition the "Driving style" button for the bicycle profile disappear until the next app start of OsmAnd. **This is caused by calculating any route using a (also a saved) GPX track**. For more clarity, here is how to generally reproduce:

**(1)** Fresh start of OsmAnd, select a (recorded, etc.) GPX track which you have displayed displayed on the map, on its track card tap the 'Directions' button.
**(2)** Route will calculate for the car mode, assuming you have not changed that default. Now change to the ped or bicycle mode.

The button I circled on the screenshot below **will not be there, and will never show up for any subsequent route configuration**, until you kill the OsmAnd session and restart.

But if instead, after the fresh app start and before step (1), you first calculate a 'normal' (non GPX-file-based) route, i.e. simply to a selected destination point, e.g. for the ped profile, the route menu for the ped profile seems correctly initialized for the rest of the OsmAnd session, and it will always display formerly missing elevation option button, even for a subsequent (saved or not-saved) GPX-based route. (Same can be done with the bicycle profile.)

For the routes where the time estimate was so far off, this is also fixed for any profile for which you first calculate "a normal route" before trying "route-based" route.

So there is something wrong with the profile selection/setup if the first route you calculate is GPX-based...
![Untitled](https://user-images.githubusercontent.com/845735/190632197-4fb399b2-5bba-47fd-9eb0-891a39139c2a.jpg) ![Untitled2](https://user-images.githubusercontent.com/845735/190632205-7eae1dbf-8e2c-4473-90e6-0073f7e0e0b4.jpg)

**Comment by ivanPyrohivskyi** [comment]
Cycling route - Cycling profile

**Comment by sonora** [comment]
PS: To reproduce you have to follow the calculated route. As soon as you have deviated, the route is recalculated and displayed completely again.

**Comment by sonora** [comment]
Let me please re-open, that fix still has glitches:

- What is does fix ok is that now there is no "permanent effect" anymore, i.e. if after a gpx-related route calculation you dismiss everything and later start a destination-related (i.e. no gpx involved) route, you will for all transport modes see the correct options again,
  - both the shortcut buttons (in the line "Options") on the route setup dialog,
  - as well as the respective profile-specific entries on the Route Settings screen.
- But whenever you perform a gpx-related route, you will now **in no case** see the correct transport-mode related options, no shortcut buttons on the route setup dialog, and no profile-specific options on the Route settings screen.

For comparison, here four screenshots (as an example for pedestrian routing), two each for destination-related and gpx-related route setups:

![Screenshot_20221109-145030_OsmAnd~](https://user-images.githubusercontent.com/845735/200850077-5141acaf-0d0b-4d4e-a41e-d197374d32a8.jpg) ![Screenshot_20221109-144952_OsmAnd~](https://user-images.githubusercontent.com/845735/200850131-45151597-b001-473d-a413-a81baddf2e45.jpg)
![Screenshot_20221109-145033_OsmAnd~](https://user-images.githubusercontent.com/845735/200850093-1ca12565-0037-4274-97cc-5fe8f3a57eb1.jpg) ![Screenshot_20221109-145004_OsmAnd~](https://user-images.githubusercontent.com/845735/200850148-49e01960-bfbe-4f9a-b1ff-547833d3be5e.jpg)

**Comment by vshcherb** [comment]
> But whenever you perform a gpx-related route, you will now in no case see the correct transport-mode related options,

That's correct options couldn't be used in GPX. We leave them in Rotue Parameters in Navigation Settings so we hide them on purpose. Even though these options might be useful for start / end segment but for most of the users it's not needed and it's very confusing.

**Comment by sonora** [comment]
That's not good ... a standard use case is this: I receive a gpx track from a friend, like for a mountain trip, and I want to find the time it would take to hike: I tap on the track in OsmAnd, then tap the "Navigation" button on its context menu, then the pedestrian icon. I do of course want to have the option for OsmAnd to e.g. use the elevation profile correction we have in our pedestrian navigation, **why not?**

I guess it's more confusing for the user if they see different options in gpx-based and destination-based routing than if they see the same?

Similarly:
- We hide the "sound" button for gpx-based routes, why?
- We hide the riding style and elevation options for bicyclists

**Comment by vshcherb** [comment]
@sonora Yes, I understand but it's not possible. If you switch profile pedestrian <-> cycling the route line doesn't change at all. 

Yes, we do have first / last segment navigation but mostly it's not used by users. 

I think as we hide toggles it's a good indication that's GPX and you shouldn't expect that enabling / disabling toggles will affect GPX route.

**Comment by sonora** [comment]
Ah, you are right... my oversight, I guess: You currently have to "_Attach to roads_" first to get e.g the actual hiking time **including elevation effects**, and then for the "Attach to roads" action these options come into play.

It's a bit weird because you can e.g. already view both the track length and the track elevation profile without "Attach to roads" being necessary, and you see a (flat ground) hiking time. And that data alone suffices to calculate also the hiking time including the elevation impact, but we don't offer to do it here. That what bugs me bit... have to think about it.

**Comment by sonora** [comment]
I think your points are ok. Only comments I have is we are not quite consistent:
- The still available options "Avoid roads..." and "Consider temporary limitations" equally make no sense and should be blanked out for gpx-routing?
- The option "Use elevation data" for pedestrians (and I think slso for bicycle - to be checked) could be offered, because its impact on the travel time solely depends on track distance, accumulated uphill, and default speed, which are all available without the user first having to attach  the tracks to roads? That would fix the use case to quickly check correct trip times for tracks representing mountain hikes (or bike trips).

**Comment by sonora** [comment]
Plus: Here's another (smaller) related glitch:
- Fresh app start
- On the map, tap on a displayed recorded track
- On its context menu, tap the navigation button
- Tap "Attach to roads"
- Pick either the bicycle or the walking profile.
- When finished, hit Apply

For the very profile you had selected for "Attach to roads", the profile-specific options and option buttons are **hidden** on the subsequent route config dialog. But if there you then switch to the **other** (bicycle or pedestrian) profile **not** originally selected for "Attach to roads", its profile-specific buttons and options are displayed ok and functional (as it should be, because now it makes sense).

**Comment by sonora** [comment]
PS2: Commit https://github.com/osmandapp/OsmAnd/commit/f10e9d05f0023530440c0bac8538e733d5e7a2e1 re-instates the originally described bug that the shortcut option buttons disappear for all subsequent destination-related routes if the first time in the current app session that profile was used for a gpx-based route.

This may also cause rhe new "smaller glitch" after "Attach to roads" I describe in the comment immediately above.

**Comment by Dima-1** [comment]
Another one try to fix https://github.com/osmandapp/OsmAnd/commit/88a2ee6721766ffd2c4a64a313e1a5a11ab9ef5f

**Comment by sonora** [comment]
Well, this is the best we have had so far, thanks!   Testing summary:

 - [x] **(1)** Route option buttons and/or route-specific config options disappear permanently if the app session's first route calculation  was gpx-based _- FIXED_
 - [x] **(2)** Mute button was missing on route config menu in gpx-based case  _- FIXED_
 - [ ] **(3)** _Optional_ - "Use elevation data" route option+button for pedestrian is missing on route config menu in gpx-based case. Could actually offer that (to determine better travel time), as it does not depend on attaching to roads, just elevation gain which is known.
 - [ ] **(4)** Start a gpx-based route calculation. When done, select "Attach to roads". After that is finished, on the subsequently popping up route config screen all mode-specific route config _buttons_ are still missing (but under _options_ they are correctly available. (I think making them available after "Attach to roads" is correct, because we have left the gpx-based case.)

**Comment by vshcherb** [comment]
@Zirochkabila please check and if possible attach video

**Comment by vshcherb** [comment]
@sonora 
(3) Optional - true deserves a separate task cause it requires rethinking and explaining in which way this setting affects the route.

**Comment by sonora** [comment]
**Regarding (3)**: You correctly pointed me the fact ;) that **nothing** you do on the "route config" screen for the gpx-based case affects the route path in any fashion. The effect is solely on the travel time we project and display (a useful feature)
- such as when tapping the icons car/ped/bicycle in the top row,
- that's why I thought we could also offer to account for the elevation effect here.

**Regarding (4)**, here it's different, needs fixing, I think: Because here we calculate a new route and use the gpx only to define the traction, analog to having many interim destinations. Here I think we need to offer all options and buttons again, and there they do affect the route path, of course.

**Comment by Zirochkabila** [comment]
OsmAnd~ 4.3.0#13943mqta, released: 2022-11-12
[telegram-cloud-document-2-5440663441549304234.mp4.zip](https://github.com/osmandapp/OsmAnd/files/9995752/telegram-cloud-document-2-5440663441549304234.mp4.zip)

**Comment by vshcherb** [comment]
As I see the left issues are minor and don't influence much on result.

Regarding 3 and 4 I tend to disagree Attach to roads still GPX navigation, so options should not be available, especially in quick access.

**Comment by sonora** [comment]
I am fine to leave as it is now for the upcoming release, but I hope we could revisit then:

The entire issue could fairly simply be avoided by 
 - [x] hiding the quick access buttons for all gpx-related route configuration -_DONE_
 - [ ] but not bothering to hide any options on the associated route "Options" ("Settings") screen at all.

Regarding the second point, we currently bother to hide some, namely all profile- specific options like "Use elevation data", even though they have a partial use without being able to change the track geometry, while we never hide other without changing track geometry clearly useless ones, like the profile-overarching "Avoid roads..." and "Considet temporary limitations".

Not hiding any menu options ever, i.e. silently leaving even 'effectless' options present and not just some, would be more consistent and require less code?

**Comment by Zirochkabila** [comment]
OsmAnd~ 4.3.0#13945mqta, released: 2022-11-14
86 maps - 10.1 GB - **Not Reproduced**

On fresh start, parameters are not available in quick access
[telegram-cloud-document-2-5445049582411062701.mp4.zip](https://github.com/osmandapp/OsmAnd/files/10002703/telegram-cloud-document-2-5445049582411062701.mp4.zip)

**Comment by sonora** [comment]
@Zirochkabila The profile-specific parameters are now always correctly available in destination-based routing, and never in gpx-based routing. That portion is solved, as I wrote above.

What is not solved is that in gpx-based routing we
- always show _profile-overarching_ settings like "Avoid roads" and "Consider temporary limitations", in your new video at 0:32, 0:36, and 1:20
- but show the _profile-specific_ settings "Riding style" and "Use elevation data" only after "Attach to the roads" (at 1:20). 

While when the track geometry mustn't change things like "Riding style" are similarly useless than "Avid roads", the opion "Use elevation data" would actually useful to have in all cases, because at least for hiking it delivers enhanced hiking times (even without changing track geometry).

**Comment by vshcherb** [comment]
It might be confusing but it's definitely not as important as other 4.3 release issues for now.

**Comment by sonora** [comment]
If you agree with my solution proposal to
- always hide the related _quick access buttons_ for gpx-based routing, 
- but always show all of each profile's _menu entries on the route settings screen_, I could make a small PR.

**Comment by sonora** [comment]
- The UI change is simple: Just remove (without substitution) `private boolean isFollowNonApproximatedGpxTrack()` and the 2 references to it. (Already successfully tested.)
- But I just notice that for the projected trip time for non-approximated gpx files we do not seem to (successfully) traverse `calculateTimeSpeed(...)`, where our Naismith/Scarf code is hosted: Our projected trip time for non-approximated gpx is always  without elevation effect, even if the setting is on. I may look into that, but it's not straight forward.
